orangecontrib/spectroscopy/widgets/owspectra.py

```diff
@@ -1009,6 +1062,101 @@ def grid_apply(self):
+                peaks, _ = find_peaks(single_spcetra)
+                max_value.append(np.amax(single_spcetra[peaks]))
+                # method for finding maxes without interference from extra data labels
+            self.maximum_point = np.amax(max_value)
+        if self.minimum_point is None:
+            self.minimum_point = np.amin(self.data)
+        if self.start_point is None:
+            self.start_point = round(np.median(getx(self.data)), 2)
+
+    def peak_apply(self):
+        if self.viewtype == INDIVIDUAL:
+            Label_line = pl.PeakLine()
+            Label_line.setMovable(True)
+            Label_line.setPos(self.start_point)
+            Label_line.label.setText(str(round(self.start_point), 3))
+            Label_line.setPen(pg.mkPen(color=QColor(Qt.black), width=2, style=Qt.DotLine))
+            Label_line.setSpan(mn=self.minimum_point, mx=self.maximum_point)
+            Label_line.label.setColor(color=QColor(Qt.black))
+            Label_line.label.setPosition(1)
+            Label_line.label.setMovable(True)
+            self.plot.addItem(Label_line)
+            Label_line.updateLabel()
+
+    def peak_apply_auto(self, prominence, minHeight, maxHeight, line_overlap):
+        x_axis = getx(self.data)
+        data = np.array(self.data)
+        peak = []
+        if np.shape(data) != (len(data),):
+            for i, val in enumerate(data):
+                single_spcetra = data[i]
+                peaks, _ = find_peaks(single_spcetra, height=([minHeight,
+                                                               maxHeight]), prominence=prominence)
+                peaks = x_axis[peaks]  # array with all locations of peaks
+                for z, val in enumerate(peaks):
+                    peak.append(peaks[z])
+            used_values = []
+            count = []
+            if line_overlap is None:
+                line_overlap = 0
+            for i, val in enumerate(peak):
+                if peak[i] not in used_values and used_values:
+                    smallest_difference = np.min(abs(np.array(used_values) - peak[i]))
+                    # comparing values to remove nearby duplicates
+                    if smallest_difference >= line_overlap:
+                        used_values.append(peak[i])
+                        count.append(1)
+                elif not used_values:
+                    used_values.append(peak[i])
+                    count.append(1)
+                    # used for first value computation as it isn't in the array
+                else:
+                    count.append(None)
+            peak_locations = []
+            for j, val in enumerate(count):
+                if count[j] is not None and count[j] == 1:
+                    peak_locations.append(peak[j])
+                    # find any peaks which aren't duplicates(None) or nearby mislabels
+            for k, val in enumerate(peak_locations):
+                Label_line = pl.PeakLine()
```

Comment:
Creating new peaks should be a function, because now, the same code pattern repeats 3 times. Perhaps peak_apply could the function if it was given proper arguments.